### PR TITLE
feat: flat config, eslint plugin, and nuxt module

### LINKED ISSUE #260: Vulnerability to track resulting from eslint-plugin-import dependency
In case there is a major semver in the eslint-plugin-import I wanted to raise this so we can track the progress and ensure that this dependency vulnerability is resolved.

See: https://github.com/import-js/eslint-plugin-import/issues/2632

The result is that we are using a version of json5 with a high vulnerability. <2.2.2

### LINKED ISSUE #288: Clarify README
The README does not currently explain the difference between `@nuxt/eslint-config` and `@nuxtjs/eslint-config`.

### LINKED ISSUE #298: migrate to typescript-eslint v6
https://typescript-eslint.io/blog/announcing-typescript-eslint-v6/

### LINKED ISSUE #303: Roadmap
The aim for this repo is to be a single source for first-class Nuxt integration with eslint (configuration + plugin(s)). We currently have a few issues.

* `@nuxt/eslint-config` is rebuilt for Nuxt 3 support without much opinion and is our preferred choice
* we still have legacy opinionated `@nuxtjs/eslint-config` and `@nuxtjs/eslint-config-typescript` configurations that are in use
* https://github.com/nuxt/eslint-plugin-nuxt is not yet updated for Nuxt 3

We'll use this issue to track key changes/tasks that need to take place. See also https://github.com/nuxt/eslint-config/issues/240.

```[tasklist]
### Tasks
- [ ] #298
- [ ] port eslint-plugin-nuxt to nuxt 3 and move here
- [ ] rename repo to nuxt/eslint
- [ ] add more extensive tests for this repo
- [ ] export several presets for less -> more opinionated linting
- [ ] deprecate `@nuxtjs/eslint-config` and `@nuxtjs/eslint-config-typescript` and remove from README
```

### LINKED ISSUE #311: Remove reference to `@nuxtjs/eslint-config` from the README
It is not clear that `@nuxt/eslint-config` is for Nuxt 3 while `@nuxtjs/eslint-config` is for Nuxt 2.
Either make it clear in the README or remove reference to `@nuxtjs/eslint-config` as it can cause confusion.
I accidentally installed the wrong one in my Nuxt 3 project.

### LINKED ISSUE #329: Error: The template root requires exactly one element. (vue/no-multiple-template-root)
I'm currently working on a Nuxt.js project and I’ve stumbled upon a problem where a page, which has already had its page transition set to false using definePageMeta, is not being recognized by ESLint.

here's my eslint config:
```module.exports = {
	root: true,
	env: {
		browser: true,
		node: true,
	},
	extends: ['@nuxt/eslint-config', 'plugin:prettier/recommended'],
	rules: {
		'no-console': 'off',
		'@typescript-eslint/no-unused-vars': [
			'error',
			{ args: 'after-used', argsIgnorePattern: '^_', varsIgnorePattern: '^_' },
		],
		'vue/multi-word-component-names': 0,
		'vue/require-default-prop': 'off',
		'vue/html-self-closing': ['error', { html: { void: 'always' } }],
		'vue/no-setup-props-destructure': 'off',
		'prefer-const': ['error', { destructuring: 'all' }],
	},
}

### LINKED ISSUE #330: `typescript-eslint` plugin version confilict.
Here is my eslint config:
```
{
  extends: [
    'eslint:recommended',
    'plugin:@typescript-eslint/recommended',
    'plugin:vue/vue3-recommended',
    '@nuxt/eslint-config',
    'prettier'
  ],
  parser: 'vue-eslint-parser',
  parserOptions: {
    parser: '@typescript-eslint/parser',
    ecmaVersion: 'latest',
    sourceType: 'module'
  },
  plugins: ['@typescript-eslint'],
  root: true
}
```
At first I just thought `@nuxt/eslint-config` was config-only for eslint, but when I upgrade `typescript-eslint` to v7, the eslint started to report error:
```
ESLint couldn't determine the plugin "@typescript-eslint" uniquely.
```
I just found `@nuxt/eslint-config` contained `typescript-eslint@6.x`, what should I do if I want to use both `typescript-eslint@v7` and `@nuxt/eslint-config`?

### LINKED ISSUE #331: Typescript Not Supported Warning
On every lint, `@typescript-eslint/typescript-estree` a dependancy of `@typescript-eslint/parser` and `@typescript-eslint/eslint-plugin` is displaying a warning when running TS >= 5.3.0:

Screenshot 2024-03-02 at 07 57 33

Likely upgrading the typescript-eslint suite to v7 will fix this.

### PR DESCRIPTION
**Pull request by antfu**
This PR merges the experiments in https://github.com/antfu/nuxt-module-eslint-config and made changes:

- `@nuxt/eslint-config`
  - Now exposes flat config sub-entry `@nuxt/eslint-config/flat`
- `@nuxt/eslint-plugin` 
  - New package, with just one rule for `process.env.dev` atm. We could add more later
- `@nuxt/eslint`
  - New Nuxt module, that generates `.nuxt/eslint.config.mjs` based on project structure and user configuration.

After this PR, we could rename the repo to `nuxt/eslint`, as the central place for hosting eslint related Nuxt integrations

## Issues

- resolve #303

#### Features

- close #329
- close https://github.com/nuxt-modules/eslint/issues/115
- close https://github.com/nuxt-modules/eslint/issues/123

#### Docs

- close #288 
- close #311 
- close https://github.com/nuxt-modules/eslint/issues/114

#### Upgrades

- close #330 
- close #331 
- close #298 
- close #260 

## Open Questions

- Q1. Should we introduce a standard docs site for all these packages? We could have an FAQ page to explain the difference between all these packages and the legacy ones.
- Q2. What should we do with https://github.com/nuxt-modules/eslint? I personally would be against running eslint along with the dev server (the idea is similar to not running type check in dev). The inline ESLint configuration for the module might also be misaligned with the configuration run in the CLI, which could cause confusion. `eslint.config.js` should be the single source of truth IMO. I would suggest we explain its downside in the README and soft-deprecate that module.

### CODE REVIEW

**Comment by Atinux** [comment]
Amazing work 🚀 

Question 1: do you plan to write a docs site like https://eslint.vuejs.org/ ?

Of adding more content in https://nuxt.com/docs/guide/concepts/code-style could do the work?

Q2: 100% agree with you, you have my go.

**Comment by antfu** [comment]
I won't mind it being `eslint.nuxt.com` or a section in `nuxt.com` - it's just that we need structural docs with a bit rich interactivity (good to have) to present different packages/concepts/rules in a nice way. If being `nuxt.com`, I guess we could configure another remote source so that the docs can still co-located inside this repo?

About Q2, I asked on Twitter, some comments for reference: https://twitter.com/antfu7/status/1767279700148174971

/cc @ricardogobbosouza in case you might have more insights about it. I personally wouldn't mind to have the module for ppl who need that capability (quicker feedback, IDE agnostic etc.), but I feel it shouldn't be the **default eslint module**. If want, we could also merge it into this repo, but with a different, more explicit name to indicate its approach. What do you think?

**Comment by ricardogobbosouza** [comment]
Hi @antfu 

Q1: I believe that a website `eslint.nuxt.com` is the best option and explains each package and module.
I would also put everything in a single `nuxt/eslint` repository

Q2: The module https://github.com/nuxt-modules/eslint is very useful, especially when we work as a team and not everyone has the IDE configured correctly, as mentioned by https://twitter.com/MatteoRigoni/status/ 1767466485834715328
I personally prefer to run alongside the development server.

In the `nuxt/eslint` repository, in my opinion, there would be packages and one for documentation:

- @nuxt/eslint-config
  -  Now exposes flat config sub-entry @nuxt/eslint-config/flat
- @nuxt/eslint-plugin
  - New package, with just one rule for process.env.dev atm. We could add more later
- @nuxt/eslint
  - New Nuxt module, that generates .nuxt/eslint.config.mjs based on project structure and user configuration
- @nuxt/eslint-moule
  - The module that runs eslint on the development server

File: packages/eslint-plugin/src/index.ts [hunk 1]
```diff
@@ -0,0 +1,9 @@
+import type { ESLint } from 'eslint'
+import rules from './rules'
+
+export default {
+  meta: {
+    name: 'nuxt-module-eslint-config/plugin',
```
Comment (danielroe) [inline packages/eslint-plugin/src/index.ts:6]: we'd need to update these names, right?

**Comment by antfu** [comment]
@ricardogobbosouza Thanks for the insights! Now I see the situation better, and yes, I think the dev server check makes sense in those cases. Tho I still think in a way it shouldn't be the default option, especially for beginners, and also I am a bit worried about the module name `@nuxt/eslint` and `@nuxtjs/eslint-module` would be confusing and not explicit enough.

What do you think about the idea that we join the force merge them together as a single module, and we could have an explicit configuration like:

```ts
export default defineNuxtConfig({
  modules: ['@nuxt/eslint'],
  eslint: {
    checker: true // or an options object 
  }
})
```

(ofc, a better name is needed, I suppose it also blocks the build process so I didn't use `devChecker`)

So that we could even have a build time only checker only if ppl want:

```ts
export default defineNuxtConfig({
  modules: ['@nuxt/eslint'],
  eslint: {
    checker: {
      dev: false,
      build: true
    }
  }
})
```

WDYT?

**Comment by ricardogobbosouza** [comment]
I agree, in merging into a single module

> @ricardogobbosouza Thanks for the insights! Now I see the situation better, and yes, I think the dev server check makes sense in those cases. Tho I still think in a way it shouldn't be the default option, especially for beginners, and also I am a bit worried about the module name `@nuxt/eslint` and `@nuxtjs/eslint-module` would be confusing and not explicit enough.
> 
> What do you think about the idea that we join the force merge them together as a single module, and we could have an explicit configuration like:
> 
> ```ts
> export default defineNuxtConfig({
>   modules: ['@nuxt/eslint'],
>   eslint: {
>     checker: true // or an options object 
>   }
> })
> ```
> 
> (ofc, a better name is needed, I suppose it also blocks the build process so I didn't use `devChecker`)
> 
> So that we could even have a build time only checker only if ppl want:
> 
> ```ts
> export default defineNuxtConfig({
>   modules: ['@nuxt/eslint'],
>   eslint: {
>     checker: {
>       dev: false,
>       build: true
>     }
>   }
> })
> ```
> 
> WDYT?

I agree to merge them together as a single module

**Review by danielroe — APPROVED** [review]
Phenomenal work @antfu. I'm really happy with this direction - this will save all of us so much time and effort.

I'm sure you don't need permission, but do feel free to merge and iterate at will. ❤️

**Comment by antfu** [comment]
Ok, this should be good to go. After this gets merged,

- [ ] Rename this repo to `nuxt/eslint`
- [ ] Setup eslint.nuxt.com
- [ ] Release alpha/beta versions for the configs and modules, and iterate
- [ ] Add messages to the README of https://github.com/nuxt/eslint-plugin-nuxt, https://github.com/nuxt-modules/eslint
- [ ] Clean up issues

/cc @ricardogobbosouza If you got a chance to take a look, I have migrated the implementation of https://github.com/nuxt-modules/eslint. This is under an opt-in `checker` option. I also added some basic docs and migration guides - let me know what you think. I also took the https://github.com/nuxt-modules/eslint/pull/124 from @ModyQyW to support flat config btw.

File: packages/module/src/modules/checker.ts [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+
+  const options: CheckerOptions = {
+    cache: true,
+    include: [`${nuxt.options.srcDir}/**/*.{js,jsx,ts,tsx,vue}`],
+    exclude: ['**/node_modules/**', nuxt.options.buildDir],
+    lintOnStart: true,
+    emitWarning: true,
+    emitError: true,
+    failOnWarning: false,
+    failOnError: false,
```
Comment (ricardogobbosouza) [inline packages/module/src/modules/checker.ts:30]: I would remove `failOnWarning` and `failOnWarning` only works with `webpack`

File: packages/module/src/modules/checker.ts [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+  // TODO: maybe support build mode later on
+  if (!nuxt.options.dev) {
+    return
+  }
+
+  const options: CheckerOptions = {
+    cache: true,
+    include: [`${nuxt.options.srcDir}/**/*.{js,jsx,ts,tsx,vue}`],
+    exclude: ['**/node_modules/**', nuxt.options.buildDir],
+    lintOnStart: true,
```
Comment (ricardogobbosouza) [inline packages/module/src/modules/checker.ts:26]: Add `formatter: 'stylish'`

File: packages/module/src/modules/checker.ts [hunk 4]
```diff
@@ -0,0 +1,92 @@
...
+      logger.warn('Please restart the Nuxt server to apply changes or upgrade to latest Nuxt for automatic restart.')
+    })
+    nuxt.hook('close', () => watcher.close())
+  }
+
+  if (nuxt.options.builder === '@nuxt/vite-builder') {
+    const vitePluginEslint = await import('vite-plugin-eslint2').then(m => m.default)
+    addVitePlugin(() => {
+      const viteOptions = {
+        linkInWorker: true,
```
Comment (ricardogobbosouza) [inline packages/module/src/modules/checker.ts:61]: Update `linkInWorker` to `lintInWorker`

**Review by ricardogobbosouza — CHANGES_REQUESTED** [review]
Need to add support for flat config in `webpack`

```js
if ( isUsingFlatConfig) {
  options.configType = 'flat'
  options.eslintPath = 'eslint/use-at-your-own-risk'
  options.overrideConfigFile: flatConfigFiles.filter(file => existsSync(resolve(nuxt.options.rootDir, file)))[0]
}

```js
    configType: 'flat',
    eslintPath: 'eslint/use-at-your-own-risk',
    overrideConfigFile: 'flat-config.js',
  })],
  // ...
};

**Comment by ricardogobbosouza** [comment]
I can help you with some things, if you think it's necessary.

- I will update `https://github.com/nuxt-modules/eslint`
- Respond to some open issues.
- Continue to support `@nuxtjs/eslint-module` until we both feel that `@nuxt/eslint` is mature enough to be a straightforward, bug-free migration (perhaps we could indicate this with a v1 release?)

**Comment by antfu** [comment]
@ricardogobbosouza Yeah, I think it makes sense to keep supporting both for a while. For the next step, we will do some alpha releases to make sure everything works as expected.

**Comment by leosvelperez** [comment]
@antfu, thanks for this! Could you split the change to update the `typescript-eslint/*` packages into a separate PR that could be released quicker than all the changes in this PR? This PR is more significant in scope and would take longer to release. Extracting the upgrade of the `typescript-eslint/*` packages would allow for addressing the associated issues more quickly.

**Comment by antfu** [comment]
This PR will be merged soon, the basic `eslintrc` remains unchanged just with deps upgrades - you should be able to upgrade directly

**Comment by leosvelperez** [comment]
Will a stable version be released soon? From [the conversation](https://github.com/nuxt/eslint-config/pull/332#issuecomment-2005187037) I thought it would need some alpha/beta versions first.

**Comment by antfu** [comment]
I suppose it would take around a week - but you can use beta version if you are hurry to upgrade `typescript-eslint`

**Comment by leosvelperez** [comment]
For context, I'm updating Nx to support TS 5.4 (needed for Angular 17.3), TypeScript ESLint ^7.2.0, but it currently conflicts with our Nuxt support. We try to avoid generating projects using pre-release versions, so batching the TypeScript ESLint change (which seems a small change) together with a bigger feature set prevents using a stable version for other tooling.

I'd be happy to send a separate PR with just that upgrade if you agree to split it. Otherwise, we're blocked or have to generate user projects with a pre-release version.

**Comment by antfu** [comment]
I understand the situation - but you know, you can always use the package manager's resolutions to force an upgrade of your dependencies locally, use our beta release (which doesn't change anything for you), or wait a week for stable, or do local patch - there are multiple workarounds you can do. We as open-source projects, couldn't really release every different combination for every moving part to make it work for every situation - keeping everything the latest is the most reasonable path we could take.

As this discussion is going off-topic, if you want to continue, I'd suggest you to create a separate issue.

**Pull request merged**